src/mrinufft/operators/autodiff.py

```diff
@@ -127,15 +135,59 @@ def __init__(self, nufft_op, wrt_data=True, wrt_traj=False):
             # used for update also.
             self._samples_torch = torch.Tensor(self.nufft_op.samples)
             self._samples_torch.requires_grad = True
+        self.batch_size = batch_size
+        self.paired_batch_mode = batch_size is not None
 
-    def op(self, x):
+    def op(self, x, smaps=None):
         r"""Compute the forward image -> k-space."""
+        if self.paired_batch_mode:
+            return self.op_batched(x, smaps)
         return _NUFFT_OP.apply(x, self.samples, self.nufft_op)
 
-    def adj_op(self, kspace):
+    def adj_op(self, kspace, smaps=None):
         r"""Compute the adjoint k-space -> image."""
+        if self.paired_batch_mode:
+            return self.adj_op_batched(kspace, smaps)
         return _NUFFT_ADJOP.apply(kspace, self.samples, self.nufft_op)
 
+    def op_batched(self, batched_imgs, batched_smaps):
+        """Compute the forward batched_imgs -> batched_kspace."""
+        # Each batch element independently calls NUFFT_OP.apply(...).
+        # The NUFFT operator is stored via ctx.nufft_op, which already saves both
+        # kspace_loc and smaps as attributes. Since ctx is isolated per element,
+        # the correct smaps are used during backpropagation.
+        self._check_input_shape(imgs=batched_imgs)
+        self._check_input_shape(imgs=batched_smaps)
+        batched_kspace = []
+        for i in range(self.batch_size):
+            try:
+                # update smaps for proper backward computation
+                self.nufft_op.smaps = batched_smaps[i]
+                batched_kspace.append(
+                    _NUFFT_OP.apply(batched_imgs[i], self.samples, self.nufft_op)
+                )
+            except Exception as e:
+                raise RuntimeError(
+                    f"Failed at batch index {i+1}"
+                ) from e  # For an easier debugging
+        return torch.stack(batched_kspace, dim=0)
+
+    def adj_op_batched(self, batched_kspace, batched_smaps):
+        """Compute the adjoint batched_kspace -> batched_imgs."""
+        # NUFFT op is saved per batch element in ctx to ensure correct backpropagation.
+        self._check_input_shape(ksps=batched_kspace)
+        self._check_input_shape(imgs=batched_smaps)
+        batched_imgs = []
+        for i in range(self.batch_size):
+            try:
+                self.nufft_op.smaps = batched_smaps[i]
```

Comment:
Careful, you could be doing a lot of copies and throttling. But I am okay with it as alternative is just to make many NUFFT ops, which I expect user to make in any case.